Feature/token validation and authorization

andreibogus opened the pull request:
This is implementation of the validation and verification of signature of the JWT tokens (SI and Access)
Closes: #244

aleksandra-bel committed:
feat: create initial class

aleksandra-bel committed:
feat: create initial classes with validation

aleksandra-bel committed:
chore: add unit tests

aleksandra-bel committed:
chore: refactor if clauses

aleksandra-bel committed:
feat: adding draft for integration test, refactoring

andreibogus committed:
feat: add JWT verification and extend tests

andreibogus committed:
feat: add JWT verification and extend tests

DominikPinsel commented on src/main/java/org/eclipse/tractusx/managedidentitywallets/utils/TokenValidationUtils.java:
```diff
@@ -0,0 +1,127 @@
+            }
+        }
+        return Optional.empty();
+    }
```
In my opinion, using the Optional object to convey errors to the caller might not be the best approach.

Instead, it could be more beneficial to return a simple boolean value for each check to the caller. I suggest splitting the methods that test various conditions The caller could then combine the results of these methods into a single object, like the following:

```json
{
  "isValid": false,
  "errors": [
    "TOKEN_EXPIRED",
    "AUDIENCE_NOT_DID",
    "NONCE_MISSING"
  ]
}
```

This approach enhances machine readability. If we aim to provide human-readable messages for the caller, I suggest avoiding this process in our services, which are typically more technical in nature.

aleksandra-bel replied:
Agree, this totally makes sense. Refactored in [adda6fe](https://github.com/borisrizov-zf/managed-identity-wallet/pull/5/commits/adda6fef7ed73b4bc7849052b61c6248978dc99e) and [34d8fe5](https://github.com/borisrizov-zf/managed-identity-wallet/pull/5/commits/34d8fe564ac29958f43a441732fa791633309a7a)

DominikPinsel commented on src/test/java/org/eclipse/tractusx/managedidentitywallets/utils/TestUtils.java:
```diff
@@ -205,4 +214,39 @@ public static String getRandomBpmNumber() {
+                .audience(audience)
+                .expirationTime(new Date(expiration))
+                .claim("nonce", nonce)
+                .build();
```
Using a `Date` object directly instead of a `long` value here is less error-prone. It eliminates the need for the caller to interpret `expiration` as milliseconds since January 1, 1970. Without this clarity, other developers might implement this differently. For instance, the Tractus-X connector project utilizes _epoch seconds_ for their time numerals.

aleksandra-bel replied:
Changed method param to Date type

DominikPinsel commented on src/main/java/org/eclipse/tractusx/managedidentitywallets/service/STSTokenValidationService.java:
```diff
@@ -0,0 +1,124 @@
+        } else {
+            log.debug(errors.toString());
+            return false;
+        }
```
It seems a `boolean` value is returned, likely because this PR is in draft mode. Nevertheless, I'll share my feedback :smile:

Rather than returning a basic boolean object, it could be beneficial for the service to return a more comprehensive response object. This approach would allow for the inclusion of additional contextual information, providing clients with a deeper understanding beyond a simple "valid" or "invalid" status.

aleksandra-bel replied:
Changed this to return Result object

DominikPinsel commented on src/main/java/org/eclipse/tractusx/managedidentitywallets/utils/CompositDidResolver.java:
```diff
@@ -0,0 +1,71 @@
+        int var3 = var2.length;
+
+        for(int var4 = 0; var4 < var3; ++var4) {
+            DidResolver didResolver = var2[var4];
```
The code seems a bit unclear. Could you please rewrite it in a more readable manner? 😄

It seems like you intended to use the SSI library here. On a slightly different note, have you had any exposure to MIW/SSI topics? If not, feel free to reach out via email! 😊

aleksandra-bel replied:
Removed this unnecessary class

aleksandra-bel committed:
chore: remove unnecessary class

refactoring

removed unnecessary class

Refactoring

borisrizov-zf commented:
@aleksandra-bel @andreibogus Hi, I've noticed that the git commits are not compliant with https://www.conventionalcommits.org/en/v1.0.0/. Please keep in mind that we'll need to update those before creating the PR in the miw repository.
Almost done with the content review, and I'll add my comments shortly.

borisrizov-zf committed:
Merge pull request #249 from borisrizov-zf/feature/token-service-and-endpoint

Feature: token service and endpoint

borisrizov-zf committed:
chore(release): 0.5.0-develop.2 [skip ci]

# [0.5.0-develop.2](https://github.com/eclipse-tractusx/managed-identity-wallet/compare/v0.5.0-develop.1...v0.5.0-develop.2) (2024-02-22)

### Bug Fixes

* add missing JSON properties ([5330f64](https://github.com/eclipse-tractusx/managed-identity-wallet/commit/5330f64dd51669d4bfd337a656c6151512ad915a))

### Features

* add domain objects needed by sts infrastructure ([effb480](https://github.com/eclipse-tractusx/managed-identity-wallet/commit/effb4801b8788c9221f6a64b4a71c990d05f3a64))
* add dto objects for sts infrastructure ([103e7f2](https://github.com/eclipse-tractusx/managed-identity-wallet/commit/103e7f2100b2613a5ab82038795c9d8f3bf06d02))
* add exceptions for sts infrastructure ([b4bb9b7](https://github.com/eclipse-tractusx/managed-identity-wallet/commit/b4bb9b78be41c721e968e1a350bd22920467b8fe))
* add missing relationships to entities ([56a48eb](https://github.com/eclipse-tractusx/managed-identity-wallet/commit/56a48eb6b4eb8a6a451512d772bf4298fce8f4f4))
* add sts properties ([018d9c0](https://github.com/eclipse-tractusx/managed-identity-wallet/commit/018d9c03430ecb15a73e753837a7a838eb7cca81))
* add token endpoint happy-path test ([d6c30bf](https://github.com/eclipse-tractusx/managed-identity-wallet/commit/d6c30bff5cec4f31498c512ecf06ed8237741d30))
* extend 'miw.security' properties ([e982919](https://github.com/eclipse-tractusx/managed-identity-wallet/commit/e9829195245780605a97e1cee5d22439c2f44c64))
* implement token endpoint ([4227f3e](https://github.com/eclipse-tractusx/managed-identity-wallet/commit/4227f3e2d147f71de84e4db6c0540e1e60849078))
* use a spring validator for SecureTokenRequest ([accc0d2](https://github.com/eclipse-tractusx/managed-identity-wallet/commit/accc0d20404852035876cb7cb4ce5ce7b5da6e64))

andreibogus committed:
chore: remove unnecessary requestMatcher

andreibogus committed:
chore: remove unnecessary class

aleksandra-bel committed:
fix: fix integration test

aleksandra-bel committed:
chore: reformat and fix exception handling
